[14.0][IMP] Cetmix Tower final preparations of 14.0-dev

ivs-cetmix opened the pull request:
# Final updates before 14.0-dev is getting merged

cetmix_tower
-------------

New module for the Odoo App Store easy installation.

cetmix_tower_server
---------------------

- Files: refactor view to make it more user-friendly
- Shortcuts: add 'Note' field, refactor views, fix tests
- Add cron job to force stop "zombie" commands
- Update description

cetmix_tower_yaml
--------------------

- Shortcuts: export the "Note' field
- Code clean up
- Description update

cetmix_tower_git
--------------------

- Add "note" field
- Code clean up
- Description update

cetmix_tower_server_notify_backend
-------------------------------------

- Code clean up
- Description update

cetmix_tower_server_queue
--------------------

- Code clean up
- Description update

Task: 4509

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced automated detection and termination of long-running (zombie) commands and added a new timeout configuration for server command execution.
	- Enhanced server template duplication and streamlined server log handling.
	- Added new fields for notes in various models to provide contextual information.
	- Added a new configuration parameter for command timeout in server settings.
- **UI Improvements**
	- Redesigned form layouts with clearer grouping and updated placeholder texts for easier configuration.
	- Improved visibility and control of fields based on variable types.
- **Documentation**
	- Updated online documentation and in-app help texts with current links and streamlined guidance for installation, configuration, and usage.
	- Removed extensive sections in README and other documentation files, directing users to official documentation for comprehensive information.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ivs-cetmix committed:
[IMP] cetmix_tower_server: UI/UX updates

- Files: refactor view to make it more user-friendly
- Shortcuts: add 'Note' field, refactor views, fix tests

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: Code cleanup

Cleanup code to comply with the OCA coding guidlines.
Add the 'note' field of the Shortcuts to export.
Fix minor issues.

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_server Code cleanup

Cleanup code to comply with the OCA coding guidlines.
Fix minor issues.

Task: 4509

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_key_value.py:
```diff
@@ -34,6 +34,13 @@ class CxTowerKeyValue(models.Model):
     )
     secret_value = fields.Text(string="Secret Value")
 
+    @api.depends("server_id", "partner_id")
+    def _compute_is_global(self):
+        for secret_value in self:
+            secret_value.is_global = (
+                not secret_value.server_id and not secret_value.partner_id
+            )
+
```
Sounds reasonable, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server_log.py:
```diff
@@ -66,27 +55,22 @@ def _selection_log_type(self):
         " when server is created from a template",
     )
 
+    def _selection_log_type(self):
+        """Actions that can be run by a command.
+
+        Returns:
+            List of tuples: available options.
+        """
+        return [
+            ("command", "Command"),
+            ("file", "File"),
+        ]
+
```
Thank you, however current docstring is enough informative.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_os.py:
```diff
@@ -4,6 +4,8 @@
 
 
 class CxTowerOs(models.Model):
+    """Operating System"""
+
```
Thank you, this will be done in the next commit.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -223,6 +147,105 @@ def write(self, vals):
+        if default is None:
+            default = {}
+
+        # skip copy name because this function use in models
+        # where it field name non store
```
Noted, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_access_role_mixin.py:
```diff
@@ -59,6 +39,26 @@ def _default_manager_ids(self):
+        # Otherwise, return an empty list. Eg if created using sudo()
+        return []
+
+    def _default_manager_ids(self):
+        """
+        Default Managers for new Records.
+        """
+        # If user is manager, add them to the list
+        if self.env.user.has_group("cetmix_tower_server.group_manager"):
+            return [self.env.user.id]
+        # Otherwise, return an empty list. Eg if created using sudo()
+        return []
```
Thank you, however those functions are not meant to be a subject of frequent changes, so currently implementation is fine.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_access_role_mixin.py:
```diff
@@ -59,6 +39,26 @@ def _default_manager_ids(self):
         copy=False,
     )
 
+    def _default_user_ids(self):
+        """
+        Default Users for new Records.
+        """
+        # If user is in group_user, add them to the list
+        if self.env.user.has_group("cetmix_tower_server.group_user"):
+            return [self.env.user.id]
+        # Otherwise, return an empty list. Eg if created using sudo()
+        return []
+
```
Thank you, however those functions are not meant to be a subject of frequent changes, so currently implementation is fine.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable.py:
```diff
@@ -23,7 +23,7 @@
 
 class TowerVariable(models.Model):
-    """Cetmix Tower Variable"""
+    """Variable values"""
```
Noted, thank you! However the correct docstring should  be "Variables"

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -762,71 +677,101 @@ def test_ssh_connection(
+        """Get command used to test SSH connection
 
-        Args:
-            command (cx.tower.command): Command to render
-            path (Char): Path where to run the command.
-                Provide in case you need to override default command path
+        Returns:
+            Char: SSH command
+        """
+        command = "uname -a"
+        return command
+
```
Other command can be easily added by inheriting and overriding this function.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -762,71 +677,101 @@ def test_ssh_connection(
+        # This is needed to avoid possible access violations
+        self.check_access_rights("read")
+        self.check_access_rule("read")
 
-            rendered_code = (
-                command.render_code_custom(command.code, **variable_values)
-                if command.code
-                else False
-            )
-            rendered_path = (
-                command.render_code_custom(path, **variable_values) if path else False
+        try:
+            client = self._get_ssh_client(
+                raise_on_error=raise_on_error, timeout=timeout
             )
 
-        else:
-            rendered_code = command.code
-            rendered_path = path
+            # Disable host key verification for this connection only, to obtain the
+            # server's real host key. If a pre-configured host key is incorrect using
+            # it would cause a key mismatch error. By setting host_key to False
+            # here, we trigger AutoAddPolicy for this connection, which automatically
+            # accepts the server's actual host key.
+            client.connection.host_key = False
 
-        return {"rendered_code": rendered_code, "rendered_path": rendered_path}
+            ssh_client = client.connection.connect()
+            transport = ssh_client.get_transport()
+            remote_key = transport.get_remote_server_key()
+            host_key = remote_key.get_base64()
+            return host_key
+        except Exception as e:
+            if raise_on_error:
+                raise ValidationError(
+                    _("Error retrieving host key: %(err)s", err=e)
+                ) from e
+            else:
+                return None
+
```
Automatic host acceptance is possible only if user has enabled it in UI. When enabled, a huge warning messages is shown that describes the possible consequences.

ivs-cetmix commented on cetmix_tower_server/wizards/cx_tower_server_template_create_wizard.py:
```diff
@@ -112,6 +112,27 @@ def _compute_missing_required_variables_message(self):
+        action.update(
+            {"view_mode": "form", "res_id": server.id, "views": [(False, "form")]}
+        )
+        return action
```
Looks reasonable, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_shortcut.py:
```diff
@@ -33,6 +38,7 @@ class CxTowerShortcut(models.Model):
         help="Run command using 'sudo'",
     )
     plan_id = fields.Many2one(string="Flight Plan", comodel_name="cx.tower.plan")
+    note = fields.Text()
```
There is a placeholder in the view that contains all necessary information.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_template_mixin.py:
```diff
@@ -80,6 +80,53 @@ def _compute_variable_ids(self):
+            return Template(code, trim_blocks=True).render(kwargs)
+        except jn_exceptions.UndefinedError as e:
+            raise UserError(e) from e
+
```
Code is entered by a user, which means user's responsibility to check the code before entering it.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file.py:
```diff
@@ -29,14 +29,16 @@
 
 class CxTowerFile(models.Model):
+    """Files"""
+
     _name = "cx.tower.file"
     _inherit = [
         "cx.tower.template.mixin",
         "mail.thread",
         "mail.activity.mixin",
         "cx.tower.key.mixin",
     ]
-    _description = "Cx Tower File"
+    _description = "Cetmix Tower File"
```
It is already changed:
```
    _description = "Cetmix Tower File"

```

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file.py:
```diff
@@ -214,28 +217,30 @@
                     and render_code_custom(file.code, **var_vals)
                     or file.code
                 )
+            rendered_name = (
+                var_vals
+                and file.name
+                and render_code_custom(file.name, **var_vals)
+                or file.name
+            )
+            rendered_server_dir = (
+                var_vals
+                and file.server_dir
+                and render_code_custom(file.server_dir, **var_vals)
+                or file.server_dir
+            )
```
This is fine, because "name" and "dir" are required fields.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_file.py:
```diff
@@ -438,6 +428,32 @@
             file.code_on_server = res
 
+    # -- Business logic
+    def _post_create_write(self, op_type="write"):
+        """Helper function that is called after file creation or update.
+        Use this function to implement custom hooks.
+
+        Args:
+            op_type (str, optional): Operation type. Defaults to "write".
+                Possible options:
+                    - "create"
+                    - "write"
+        """
+
+        # Pull all `auto_sync` server files
+        server_files_to_sync = self.filtered(
+            lambda file: file.auto_sync and file.source == "server"
+        )
+        if server_files_to_sync:
+            server_files_to_sync.action_pull_from_server()
+
+        # Push all `auto_sync` tower files
+        tower_files_to_sync = self.filtered(
+            lambda file: file.auto_sync and file.source == "tower"
+        )
+        if tower_files_to_sync:
+            tower_files_to_sync.action_push_to_server()
+
```
Yes, this is a planned feature to be added in the "cetmix_tower_server_queue" module.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -353,14 +323,23 @@ def unlink(self):
-        """
-        self.ensure_one()
-        return self.status and self.status == "deleting"
+    def _read(self, fields):
+        """Substitute fields based on api"""
+        super()._read(fields)
+        if not self.env.context.get("show_host_key") and (
+            "host_key" in fields or fields == []
+        ):
+            spoiler = self.env["cx.tower.key"].SECRET_VALUE_SPOILER
+            # Public user used for substitution
+            for record in self:
+                try:
+                    record._cache["host_key"] = (
+                        spoiler if record._cache["host_key"] else None
+                    )
+                except Exception:
+                    # skip SpecialValue
+                    # (e.g. for missing record or access right)
+                    pass
 
```
Yup, we are aware of it, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -762,71 +677,101 @@
+        """Get command used to test SSH connection
 
-        Args:
-            command (cx.tower.command): Command to render
-            path (Char): Path where to run the command.
-                Provide in case you need to override default command path
+        Returns:
+            Char: SSH command
+        """
+        command = "uname -a"
+        return command
+
```
This was already replied. Once again: you can inherit and override this function to add any custom command you like.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -762,71 +677,101 @@
+        # This is needed to avoid possible access violations
+        self.check_access_rights("read")
+        self.check_access_rule("read")
 
-            rendered_code = (
-                command.render_code_custom(command.code, **variable_values)
-                if command.code
-                else False
-            )
-            rendered_path = (
-                command.render_code_custom(path, **variable_values) if path else False
+        try:
+            client = self._get_ssh_client(
+                raise_on_error=raise_on_error, timeout=timeout
             )
 
-        else:
-            rendered_code = command.code
-            rendered_path = path
+            # Disable host key verification for this connection only, to obtain the
+            # server's real host key. If a pre-configured host key is incorrect using
+            # it would cause a key mismatch error. By setting host_key to False
+            # here, we trigger AutoAddPolicy for this connection, which automatically
+            # accepts the server's actual host key.
+            client.connection.host_key = False
 
-        return {"rendered_code": rendered_code, "rendered_path": rendered_path}
+            ssh_client = client.connection.connect()
+            transport = ssh_client.get_transport()
+            remote_key = transport.get_remote_server_key()
+            host_key = remote_key.get_base64()
+            return host_key
+        except Exception as e:
+            if raise_on_error:
+                raise ValidationError(
+                    _("Error retrieving host key: %(err)s", err=e)
+                ) from e
+            else:
+                return None
+
```
This was already replied before. Automatic host key acceptance is possible only if user has enabled it explicitly in the UI. When enabled, a huge warning is show with all possible consequences explained.

ivs-cetmix committed:
[IMP] cetmix_tower_git: UI/UX improvements

- Add 'Note' field to git project
- Clean up code before merging into 14.0

Task: 4509

ivs-cetmix commented on cetmix_tower_server/static/description/index.html:
```diff
@@ -369,942 +369,60 @@ <h1 class="title">Cetmix Tower Server</h1>
-different operating systems and software environments, providing a
-practical option for those looking to manage servers without getting
+for managing remote servers and applications via SSH or API calls
+directly from <a class="reference external" href="https:/odoo.com">Odoo</a>. It is designed for versatility
```
This file is generated automatically. Please remember it finally.

ivs-cetmix commented on cetmix_tower_server/static/description/index.html:
```diff
@@ -369,942 +369,60 @@ <h1 class="title">Cetmix Tower Server</h1>
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! -->
-<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Beta" src="https://img.shields.io/badge/maturity-Beta-yellow.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/licence-AGPL--3-blue.png" /></a> <a class= ...
+<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Beta" src="https://img.shields.io/badge/maturity-Beta-yellow.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/licence-AGPL--3-blue.png" /></a> <a class= ...
 <p><a class="reference external" href="http://cetmix.com/tower">Cetmix Tower</a> offers a streamlined solution
```
This file is generated automatically. Please remember it finally.

ivs-cetmix commented on cetmix_tower_server/static/description/index.html:
```diff
@@ -369,942 +369,60 @@ <h1 class="title">Cetmix Tower Server</h1>
-<li>For additional details, visit our website
-<a class="reference external" href="https://cetmix.com">cetmix.com</a></li>
-</ul>
+<p>Please refer to the <a class="reference external" href="https://cetmix.com/tower">official
+documentation</a> for detailed information and
+documentation.</p>
```
Noted, thank you

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -221,8 +149,112 @@ def write(self, vals):
+        if "reference" not in default:
+            default["reference"] = self._generate_or_fix_reference(default["name"])
+        return super().copy(default=default)
+
+    def _get_reference_pattern(self):
+        """
+        Returns the regex pattern used for validating and correcting references.
+        This allows for easy modification of the pattern in one place.
+
+        Important: pattern must be enclosed in square brackets!
+
+        Returns:
+            str: A regex pattern
+        """
+        return "[a-z0-9_]"
```
REFERENCE_PRELIMINARY_PATTERN is used for automatically populated references which are later corrected in the _generate_of_fix_reference function.

ivs-cetmix commented on cetmix_tower_git/models/cx_tower_git_project.py:
```diff
@@ -86,6 +86,7 @@ def _get_post_create_fields(self):
         depends=["git_project_file_template_rel_ids"],
         copy=False,
     )
+    note = fields.Text()
```
There is a placeholder added for this field in the view that explains what it is used for.

ivs-cetmix commented on cetmix_tower_git/models/cx_tower_git_remote.py:
```diff
@@ -141,6 +118,29 @@ def _compute_head_type(self):
+                    raise ValidationError(
+                        _("Not a valid URL. URL must start with 'https://' or 'git@'")
+                    )
+
```
Noted, thank you!

ivs-cetmix committed:
[IMP] cetmix_tower_notify_backend Code cleanup

Cleanup code to comply with the OCA coding guidlines.
Fix minor issues.

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_server_queue Code cleanup

Cleanup code to comply with the OCA coding guidlines.
Fix minor issues.

Task: 4509

ivs-cetmix commented on cetmix_tower_server/readme/USAGE.md:
```diff
@@ -1,120 +1 @@
-You can use various Cetmix Tower functions in Odoo automation. Such as server actions, automated actions or scheduled actions.
-While you can always call any public function directly most useful helper functions are located in a special abstract model "cetmix.tower".
-You can check those functions in the source code in the following file: `models/cetmix_tower.py`
+Please refer to the [official documentation](https://cetmix.com/tower) for detailed usage instructions.
```
This file is used for the automated generation of the README.rst file. So all headers are generated automatically

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server_template.py:
```diff
@@ -117,13 +113,33 @@ def _get_post_create_fields(self):
         for template in self:
             template.server_count = len(template.server_ids)
 
+    def copy(self, default=None):
+        """Duplicate the server template along with variable values and server logs."""
+        default = dict(default or {})
+
+        # Duplicate the server template itself
+        new_template = super().copy(default)
+
+        # Duplicate variable values
+        for variable_value in self.variable_value_ids:
+            variable_value.with_context(reference_mixin_skip_self=True).copy(
+                {"server_template_id": new_template.id}
+            )
+
+        # Duplicate server logs
+        for server_log in self.server_log_ids:
+            server_log.copy({"server_template_id": new_template.id})
+
+        return new_template
+
```
This is done automatically by Odoo ORM

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_template_mixin.py:
```diff
@@ -80,6 +80,53 @@ def _compute_variable_ids(self):
+        for rec in self:
+            rendered_code = self.render_code_custom(rec.code, pythonic_mode, **kwargs)
+            res.update({rec.id: rendered_code})
+
```
Noted, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -121,17 +121,50 @@ class TowerVariableValue(models.Model):
+        for rec in self:
+            rec.is_global = rec._check_is_global()
+
+    @api.depends("option_id", "variable_id.option_ids")
+    def _compute_value_char(self):
+        """
+        Compute the 'value_char' field, which holds the string representation
+        of the selected option for the variable.
+        """
+        for rec in self:
+            if rec.variable_id.option_ids and rec.option_id:
+                rec.value_char = rec.option_id.value_char
+            elif not rec.variable_id.option_ids:
+                rec.value_char = rec.value_char or ""
+                rec.option_id = None
+
```
Thank you for noticing this!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -304,7 +313,7 @@ def get_by_reference(self, reference):
         return self.browse(self._get_id_by_reference(reference))
 
-    # TODO: implement caching for this method
+    @ormcache("reference")
     def _get_id_by_reference(self, reference):
```
It is already done in both "create" and "write" functions

ivs-cetmix committed:
[FIX] Variable Values

- Fix variable value check
- Reset option_id when value_char is invalid
- Clear value_char when option_id is removed

ivs-cetmix commented on cetmix_tower_server_notify_backend/readme/CONFIGURE.md:
```diff
@@ -1 +1 @@
-No configuration is required.
+Please refer to the [official documentation](https://cetmix.com/tower) for detailed configuration instructions.
```
This file is used for generation of the README.rst file, so all headers are created automatically.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_template_mixin.py:
```diff
@@ -80,6 +80,53 @@ def _compute_variable_ids(self):
             else:
                 record.variable_ids = [(5, 0, 0)]
 
+    def render_code(self, pythonic_mode=False, **kwargs):
+        """Render record 'code' field using variables from kwargs
+        Call to render recordset of the inheriting models
+
+        Args:
+            pythonic_mode (Bool): If True, all variables in kwargs are converted to
+                                  strings and wrapped in double quotes.
+                                  Default is False.
+            **kwargs (dict): {variable: value, ...}
+        Returns:
+            dict {record_id: rendered_code, ...}
+        """
+        res = {}
+        for rec in self:
+            rendered_code = self.render_code_custom(rec.code, pythonic_mode, **kwargs)
+            res.update({rec.id: rendered_code})
+
+        return res
+
```
Looks reasonable, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_template_mixin.py:
```diff
@@ -80,6 +80,53 @@ def _compute_variable_ids(self):
+                    for key, value in kwargs.items()
+                }
+            return Template(code, trim_blocks=True).render(kwargs)
+        except jn_exceptions.UndefinedError as e:
+            raise UserError(e) from e
```
Looks reasonable, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_template_mixin.py:
```diff
@@ -80,6 +79,50 @@ def _compute_variable_ids(self):
             else:
                 record.variable_ids = [(5, 0, 0)]
 
+    def render_code(self, pythonic_mode=False, **kwargs):
+        """Render record 'code' field using variables from kwargs
+        Call to render recordset of the inheriting models
+        Args:
+            pythonic_mode (Bool): If True, all variables in kwargs are converted to
+                                  strings and wrapped in double quotes.
+                                  Default is False.
+            **kwargs (dict): {variable: value, ...}
+        Returns:
+            dict {record_id: rendered_code, ...}
+        """
+        return {
+            rec.id: self.render_code_custom(rec.code, pythonic_mode, **kwargs)
+            for rec in self
+        }
```
This is handled in the render_code_custom function

ivs-cetmix commented on cetmix_tower_yaml/README.rst:
```diff
@@ -17,155 +17,43 @@ Cetmix Tower YAML
 
-This module implements YAML format data import/export for `Cetmix
-Tower. <https://cetmix.com/tower>`__
+This module implements YAML format data import/export for `Cetmix
+Tower <https://cetmix.com/tower>`__.
 
```
This file is auto generated from template. Please remember this forever for all repos.

ivs-cetmix commented on cetmix_tower_yaml/readme/DESCRIPTION.md:
```diff
@@ -1,9 +1,3 @@
-This module implements YAML format data import/export for [Cetmix Tower.](https://cetmix.com/tower)
+This module implements YAML format data import/export for [Cetmix Tower](https://cetmix.com/tower).
```
This file is used for the README.rst generation. So all headers are generated automatically. Please remember this for all modules in all repos.

ivs-cetmix committed:
[FIX] Server Host Key

Do not check host key if IP address is not present.
This may happen when server is created from template and IP address is
assigned by a flight plan that runs after server creation.

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -221,8 +149,112 @@ def write(self, vals):
+        if "reference" in vals:
+            self.clear_caches()
+
         return super().write(vals)
```
also:
```python
    def unlink(self):
        res = super().unlink()
        self.clear_caches()
        return res
```

ivs-cetmix replied:
Noted, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -16,7 +17,7 @@ class CxTowerReferenceMixin(models.AbstractModel):
 
     # Used to check the reference before it's being fixed.
-    # THis is needed to ensure that there is at least one
+    # This is needed to ensure that there is at least one
```
Noted, thank you!

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_server_view.xml:
```diff
@@ -324,7 +321,7 @@
                                     <field name="ssh_port" />
                                     <field name="ssh_username" />
-                                    <field name="use_sudo" />
+                                    <field name="use_sudo" placeholder="No" />
```
We add No/Undefined in server template and wizard because it's not clear at the moment when wizard is launched if user doesn't want to use sudo or it will be defined later.
While when server is already created it becomes clear.

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_server_view.xml:
```diff
@@ -267,10 +267,7 @@
-                                </group>
+                                <field name="note" />
                             </form>
                         </field>
```
Yes, because user doesn't need techical information, just an explanation what this shortcut does.

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_server_template_view.xml:
```diff
@@ -177,7 +177,7 @@
                                     <field name="ssh_port" />
                                     <field name="ssh_username" />
-                                    <field name="use_sudo" />
+                                    <field name="use_sudo" placeholder="No/Undefined" />
```
We add No/Undefined in server template and wizard because it's not clear at the moment when wizard is launched if user doesn't want to use sudo or it will be defined later.
While when server is already created it becomes clear.

ivs-cetmix commented on cetmix_tower_server/wizards/cx_tower_server_template_create_wizard_view.xml:
```diff
@@ -44,7 +44,7 @@
                             attrs="{'invisible': [('ssh_auth_mode', '=', 'p')]}"
                         />
-                        <field name="use_sudo" />
+                        <field name="use_sudo" placeholder="No/Undefined" />
```
We add No/Undefined in server template and wizard because it's not clear at the moment when wizard is launched if user doesn't want to use sudo or it will be defined later.
While when server is already created it becomes clear.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -221,8 +149,120 @@ def write(self, vals):
+            default["reference"] = self._generate_or_fix_reference(default["name"])
+        return super().copy(default=default)
+
+    def _get_reference_pattern(self):
+        """
+        Returns the regex pattern used for validating and correcting references.
+        This allows for easy modification of the pattern in one place.
+
+        Important: pattern must be enclosed in square brackets!
+
+        Returns:
+            str: A regex pattern
+        """
+        return "[a-z0-9_]"
+
```
This is clear, thank you

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -186,7 +111,10 @@ def create(self, vals_list):
                 )
-        return super().create(vals_list)
+
+        res = super().create(vals_list)
+        self.clear_caches()
+        return res
```
This is clear, thank you

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_key.py:
```diff
@@ -58,6 +59,7 @@ class CxTowerKey(models.Model):
         relation="cx_tower_key_manager_rel",
     )
 
+    @api.depends("reference")
```
Sounds reasonable, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -357,6 +324,39 @@ def _inverse_is_global(self):
                     )
                 )
 
+    def _inverse_value_char(self):
+        """Set option_id based on value_char"""
+        for rec in self:
+            if rec.variable_type == "o" and (
+                not rec.option_id or rec.option_id.value_char != rec.value_char
+            ):
+                option = rec.variable_id.option_ids.filtered(
+                    lambda x, v=rec.value_char: x.value_char == v
+                )
+                rec.option_id = option and option.id
+
```
value_char is unique per option

ivs-cetmix replied:
This scenario never arises thanks to constaints

ivs-cetmix replied:
They never share the same name because of the SQL constraint

ivs-cetmix commented on cetmix_tower_git/readme/CONFIGURE.md:
```diff
@@ -1,2 +1 @@
-For information on how to configure Git Projects, please refer to the official documentation: https://cetmix.com/tower
-
+Please refer to the [official documentation](https://cetmix.com/tower) for detailed configuration instructions.
```
Once again: this file is used for the README.rst generation. All headers will be added automatically. Please remember it for all repos and NEVER suggest to add headers here.

ivs-cetmix replied:
How many times should I repeat that ALL /readme files are used for the README.rst generation and those headers are generated automatically? This is true for ALL odoo modules in ALL repos. Could you please finally remember this and stop posting the same "issue" over and over again?

ivs-cetmix replied:
Are you kidding? How much time should I repeat, that this file is used for the README.rst generation. So all headers are generated automatically. Please remember this for all Odoo modules in all repos.

ivs-cetmix replied:
Are you kidding me?
How many times should I repeat: 
All files in the /readme directory  are used for generation of the README.rst file, so all headers are created automatically. In ALL Odoo modules. In ALL repos

ivs-cetmix commented on cetmix_tower_git/readme/DESCRIPTION.md:
```diff
@@ -1,3 +1,3 @@
-This module implements Git Management tools for **Cetmix Tower**.
+This module implements Git Management functionality for [Cetmix Tower](https://cetmix.com/tower).
 
-For detailed information please refer to the official documentation: https://cetmix.com/tower
+Please refer to the [official documentation](https://cetmix.com/tower) for detailed information.
```
Once again: this file is used for the README.rst generation. All headers will be added automatically. Please remember it for all repos and NEVER suggest to add headers in this file in all repos.

ivs-cetmix replied:
How many times should I repeat that ALL /readme files are used for the README.rst generation and those headers are generated automatically? This is true for ALL odoo modules in ALL repos. Could you please finally remember this and stop posting the same "issue" over and over again?

ivs-cetmix replied:
Are you kidding? How much time should I repeat, that this file is used for the README.rst generation. So all headers are generated automatically. Please remember this for all Odoo modules in all repos.

ivs-cetmix replied:
Are you kidding me?
How many times should I repeat: 
All files in the /readme directory  are used for generation of the README.rst file, so all headers are created automatically. In ALL Odoo modules. In ALL repos

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server_template.py:
```diff
@@ -117,13 +114,33 @@ def _get_post_create_fields(self):
         for template in self:
             template.server_count = len(template.server_ids)
 
+    def copy(self, default=None):
+        """Duplicate the server template along with variable values and server logs."""
+        default = dict(default or {})
+
+        # Duplicate the server template itself
+        new_template = super().copy(default)
+
+        # Duplicate variable values
+        for variable_value in self.variable_value_ids:
+            variable_value.with_context(reference_mixin_skip_self=True).copy(
+                {"server_template_id": new_template.id}
+            )
+
+        # Duplicate server logs
+        for server_log in self.server_log_ids:
+            server_log.copy({"server_template_id": new_template.id})
+
+        return new_template
```
Server logs don't have any issues with references.

ivs-cetmix replied:
This is not a case for server logs

ivs-cetmix commented on cetmix_tower_server_notify_backend/readme/DESCRIPTION.md:
```diff
@@ -1,3 +1,5 @@
-Please check the official documentation for more information: https://cetmix.com/tower
+It requires the [web_notify](https://github.com/OCA/web/web_notify) module to be installed and configured in the Odoo instance.
+
+Please refer to the [official documentation](https://cetmix.com/tower) for detailed information.
```
How many times should I repeat that ALL /readme files are used for the README.rst generation and those headers are generated automatically? This is true for ALL odoo modules in ALL repos. Could you please finally remember this and stop posting the same "issue" over and over again?

ivs-cetmix replied:
Are you kidding? How much time should I repeat, that this file is used for the README.rst generation. So all headers are generated automatically. Please remember this for all Odoo modules in all repos.

ivs-cetmix replied:
Are you kidding me?
How many times should I repeat: 
All files in the /readme directory  are used for generation of the README.rst file, so all headers are created automatically. In ALL Odoo modules. In ALL repos

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -121,17 +121,53 @@ class TowerVariableValue(models.Model):
+
+    @api.depends("variable_id", "variable_id.access_level")
+    def _compute_access_level(self):
         """
-        Check if the value_char is valid for the variable.
+        Automatically set the access_level based on Variable access level
         """
         for rec in self:
-            if not rec.variable_id:
+            if rec.variable_id:
+                rec.access_level = rec.variable_id.access_level
+
```
this is a mandatory field, so an exception will be raised automatically

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -121,17 +121,53 @@ class TowerVariableValue(models.Model):
+            else:
+                rec.value_char = False
+
+    @api.depends("value_char")
+    def _compute_variable_ids(self):
+        """
+        Compute variable_ids based on value_char field.
+        """
+        template_mixin_obj = self.env["cx.tower.template.mixin"]
+        for record in self:
+            record.variable_ids = template_mixin_obj._prepare_variable_commands(
+                ["value_char"], force_record=record
+            )
+
```
option_id doesn't contain variables

ivs-cetmix commented on cetmix_tower_server_queue/static/description/index.html:
```diff
@@ -274,7 +275,7 @@
   margin-right: 2em }
 
-pre.code .ln { color: grey; } /* line numbers */
+pre.code .ln { color: gray; } /* line numbers */
 pre.code, code { background-color: #eeeeee }
```
How many times should I repeat that index.html is generated automatically?. Could you please finally remember this and stop posting the same "issue" over and over again?

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_template_mixin.py:
```diff
@@ -80,6 +79,50 @@ def _compute_variable_ids(self):
+            for rec in self
+        }
+
+    def render_code_custom(self, code, pythonic_mode=False, **kwargs):
+        """
+        Render custom code using variables from kwargs
+
+        This method renders a template string (code) using the variables provided
+        in kwargs. If pythonic_mode is enabled, all variables are automatically
+        converted to strings and enclosed in double quotes before rendering.
+
+        Args:
+            code (Text): code to render (eg 'some {{ custom }} text')
+            pythonic_mode (Bool): If True, all variables in kwargs are converted to
+                                  strings and wrapped in double quotes.
+                                  Default is False.
+            **kwargs (dict): {variable: value, ...}
+        Returns:
+            rendered_code (text): The resulting string after rendering the template with
+                                  the provided variables.
+        """
+        try:
+            if pythonic_mode:
+                kwargs = {
+                    key: self._make_value_pythonic(value)
+                    for key, value in kwargs.items()
+                }
+            return Template(code, trim_blocks=True).render(kwargs)
+        except Exception as e:
+            raise UserError(str(e)) from e
+
```
We don't need to distinguish those errors.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -447,140 +430,72 @@ def action_open_plan_logs(self):
-        self.ensure_one()
-        password = self.ssh_password
-        return password
-
-    def _get_ssh_key(self):
-        """Get SSH key
-        Get private key for an SSH connection
+        context = self.env.context.copy()
+        context.update(
+            {
+                "default_server_ids": self.ids,
+            }
+        )
+        return {
+            "type": "ir.actions.act_window",
+            "name": _("Run Flight Plan"),
+            "res_model": "cx.tower.plan.run.wizard",
+            "view_mode": "form",
+            "view_type": "form",
+            "target": "new",
+            "context": context,
+        }
```
Noted, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -221,8 +149,120 @@ def write(self, vals):
+        reference_pattern = self._get_reference_pattern()
+
+        if re.fullmatch(rf"{reference_pattern}+", reference_source):
+            reference = reference_source
+
+        # Fix reference if it doesn't match
+        else:
+            # Modify the pattern to be used in `sub`
+            inner_pattern = reference_pattern[1:-1]
+            reference = (
+                re.sub(
+                    rf"[^{inner_pattern}]",
+                    "",
+                    reference_source.strip().replace(" ", "_").lower(),
+                )
+                or self._get_model_generic_reference()
+            )
+
+        # Check if the same reference already exists and add a suffix if yes
+        counter = 1
+        final_reference = reference
+
+        # If exclude same records from search results
+        if self and not self.env.context.get("reference_mixin_skip_self"):
+            domain = [("id", "not in", self.ids)]
+        else:
+            domain = []
+        final_domain = expression.AND([domain, [("reference", "=", final_reference)]])
+
+        # Search all records without restrictions including archived
+        self_with_sudo_and_context = self.sudo().with_context(active_test=False)
+        while self_with_sudo_and_context.search_count(final_domain) > 0:
+            counter += 1
+            final_reference = _(f"{reference}_{counter}")
+            final_domain = expression.AND(
+                [domain, [("reference", "=", final_reference)]]
+            )
+
+        return final_reference
+
```
Noted, thank you!

ivs-cetmix commented on cetmix_tower_server/tests/test_variable.py:
```diff
@@ -39,7 +39,7 @@ def check_variable_values(self, vals, server_ids=None):
                 len(variable_line), 1, msg="Must be a single variable line"
             )
-            expected_value = val[1] or ""
+            expected_value = val[1] or False
```
Thank you for the feedback! But what exactly is wrong?

ivs-cetmix replied:
What exactly is wrong here?

ivs-cetmix commented on cetmix_tower_server_queue/static/description/index.html:
```diff
@@ -8,10 +8,11 @@
 
 /*
 :Author: David Goodger (goodger@python.org)
-:Id: $Id: html4css1.css 8954 2022-01-20 10:10:25Z milde $
+:Id: $Id: html4css1.css 9511 2024-01-13 09:50:07Z milde $
 :Copyright: This stylesheet has been placed in the public domain.
 
 Default cascading style sheet for the HTML output of Docutils.
+Despite the name, some widely supported CSS2 features are used.
 
```
Are you kidding? How much time should I repeat, that this file is generated AUTOMATICALLY. Please remember this for all Odoo modules in all repos.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -145,7 +141,7 @@ def _get_post_create_fields(self):
     use_sudo = fields.Selection(
         string="Use sudo",
         selection=[("n", "Without password"), ("p", "With password")],
-        help="Run commands using 'sudo'",
+        help="Run commands using 'sudo'. Leave empty if 'sudo' is not needed.",
         groups="cetmix_tower_server.group_manager",
     )
```
'n' and 'p' are selection values used in code and are not visible in UI

ivs-cetmix committed:
[ADD] cetmix_tower App Store Module

This module is a collection of tools for the Tower project.
It's main purpose is to be used in the Odoo App Store so users could
get all modules at once without searching for them.

Task: 4509

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -357,6 +324,39 @@ def _inverse_is_global(self):
                     )
                 )
 
+    def _inverse_value_char(self):
+        """Set option_id based on value_char"""
+        for rec in self:
+            if rec.variable_type == "o" and (
+                not rec.option_id or rec.option_id.value_char != rec.value_char
+            ):
+                option = rec.variable_id.option_ids.filtered(
+                    lambda x, v=rec.value_char: x.value_char == v
+                )
+                rec.option_id = option and option.id
+
```
filtered() uses cache with may be good from the performance point of view. Especially in this particular case when we are applying this method to a one2may field records.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -221,8 +149,120 @@ def write(self, vals):
+        reference_pattern = self._get_reference_pattern()
+
+        if re.fullmatch(rf"{reference_pattern}+", reference_source):
+            reference = reference_source
+
+        # Fix reference if it doesn't match
+        else:
+            # Modify the pattern to be used in `sub`
+            inner_pattern = reference_pattern[1:-1]
+            reference = (
+                re.sub(
+                    rf"[^{inner_pattern}]",
+                    "",
+                    reference_source.strip().replace(" ", "_").lower(),
+                )
+                or self._get_model_generic_reference()
+            )
+
+        # Check if the same reference already exists and add a suffix if yes
+        counter = 1
+        final_reference = reference
+
+        # If exclude same records from search results
+        if self and not self.env.context.get("reference_mixin_skip_self"):
+            domain = [("id", "not in", self.ids)]
+        else:
+            domain = []
+        final_domain = expression.AND([domain, [("reference", "=", final_reference)]])
+
+        # Search all records without restrictions including archived
+        self_with_sudo_and_context = self.sudo().with_context(active_test=False)
+        while self_with_sudo_and_context.search_count(final_domain) > 0:
+            counter += 1
+            final_reference = _(f"{reference}_{counter}")
+            final_domain = expression.AND(
+                [domain, [("reference", "=", final_reference)]]
+            )
+
+        return final_reference
+
```
This is not our case, however thank you for noticing this.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -186,7 +111,10 @@ def create(self, vals_list):
                     {"reference": self._generate_or_fix_reference(reference or name)}
                 )
-        return super().create(vals_list)
+
+        res = super().create(vals_list)
+        self.clear_caches()
+        return res
```
I'm aware of this. Thank you.

ivs-cetmix commented on cetmix_tower_server/readme/DESCRIPTION.md:
```diff
@@ -1,50 +1,4 @@
-[Cetmix Tower](http://cetmix.com/tower) offers a streamlined solution for managing remote servers via SSH or API calls directly from [Odoo](https:/odoo.com).
+[Cetmix Tower](http://cetmix.com/tower) offers a streamlined solution for managing remote servers and applications via SSH or API calls directly from [Odoo](https:/odoo.com).
```
Noted, thank you!

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_access_mixin.py:
```diff
@@ -9,6 +9,13 @@ class CxTowerAccessMixin(models.AbstractModel):
     _name = "cx.tower.access.mixin"
     _description = "Cetmix Tower access mixin"
 
+    access_level = fields.Selection(
+        lambda self: self._selection_access_level(),
+        default=lambda self: self._default_access_level(),
+        required=True,
+        index=True,
+    )
```
String will be generated automatically from the field name: access_level -> Access Level

ivs-cetmix commented on cetmix_tower_server/wizards/cx_tower_command_run_wizard.py:
```diff
@@ -188,6 +182,13 @@ def _compute_os_compatibility_warning(self):
                 "\n".join(warning_list) if warning_list else False
             )
 
+    @api.onchange("action", "applicability")
+    def _onchange_action(self):
+        """
+        Reset command after change action
+        """
+        self.command_id = False
```
This is not required because it is done in a dev branch

ivs-cetmix commented on cetmix_tower_yaml/readme/USAGE.md:
```diff
@@ -1,73 +1 @@
-This mode allows to export/import child records together with the parent one.
-In case any of the child fields are modified in YAML related record in Odoo will be modified using those values.
-In case related record cannot be resolved using child reference while importing data from YAML, new child record will be created in Odoo using YAML values.
+Please refer to the [official documentation](https://cetmix.com/tower) for detailed usage instructions.
```
What? Again? Is this a joke? Please check the comments regarding files in the /readme directorey

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -447,140 +430,70 @@ def action_open_plan_logs(self):
-            )
+        Open files of the current server
+        """
+        action = self.env["ir.actions.actions"]._for_xml_id(
+            "cetmix_tower_server.cx_tower_file_action"
+        )
+        action["domain"] = [("server_id", "=", self.id)]  # pylint: disable=no-member
+
+        context = self._context.copy()
+        if "context" in action and isinstance((action["context"]), str):
+            context.update(ast.literal_eval(action["context"]))
         else:
-            ssh_key = None
-        return ssh_key
+            context.update(action.get("context", {}))
 
-    def _get_connection_test_command(self):
-        """Get command used to test SSH connection
+        context.update(
+            {
+                "default_server_id": self.id,  # pylint: disable=no-member
+            }
+        )
+        action["context"] = context
+        return action
 
```
Yes, we know that. Thank you.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -221,8 +149,120 @@ def write(self, vals):
+        reference_pattern = self._get_reference_pattern()
+
+        if re.fullmatch(rf"{reference_pattern}+", reference_source):
+            reference = reference_source
+
+        # Fix reference if it doesn't match
+        else:
+            # Modify the pattern to be used in `sub`
+            inner_pattern = reference_pattern[1:-1]
+            reference = (
+                re.sub(
+                    rf"[^{inner_pattern}]",
+                    "",
+                    reference_source.strip().replace(" ", "_").lower(),
+                )
+                or self._get_model_generic_reference()
+            )
+
+        # Check if the same reference already exists and add a suffix if yes
+        counter = 1
+        final_reference = reference
+
+        # If exclude same records from search results
+        if self and not self.env.context.get("reference_mixin_skip_self"):
+            domain = [("id", "not in", self.ids)]
+        else:
+            domain = []
+        final_domain = expression.AND([domain, [("reference", "=", final_reference)]])
+
+        # Search all records without restrictions including archived
+        self_with_sudo_and_context = self.sudo().with_context(active_test=False)
+        while self_with_sudo_and_context.search_count(final_domain) > 0:
+            counter += 1
+            final_reference = _(f"{reference}_{counter}")
+            final_domain = expression.AND(
+                [domain, [("reference", "=", final_reference)]]
+            )
+
+        return final_reference
+
```
Yes we know that. Thank you.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -186,7 +111,10 @@ def create(self, vals_list):
                     {"reference": self._generate_or_fix_reference(reference or name)}
                 )
-        return super().create(vals_list)
+
+        res = super().create(vals_list)
+        self.clear_caches()
+        return res
```
Yes we know that. Thank you.

ivs-cetmix commented on setup/cetmix_tower/odoo/addons/cetmix_tower:
```diff
@@ -0,0 +1 @@
+../../../../cetmix_tower
```
Yes, because this is a Python setup file.

ivs-cetmix commented on cetmix_tower/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+# Copyright Cetmix OU
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Cetmix Tower",
+    "summary": "Odoo SAAS Server Application Management",
+    "version": "14.0.1.0.0",
+    "category": "Productivity",
+    "website": "https://cetmix.com",
+    "live_test_url": "https://cetmix.com/tower",
+    "author": "Cetmix",
+    "license": "AGPL-3",
+    "application": True,
+    "installable": True,
+    "depends": [
+        "cetmix_tower_server",
+        "cetmix_tower_server_queue",
+        "cetmix_tower_server_notify_backend",
+        "cetmix_tower_git",
+        "cetmix_tower_yaml",
+    ],
+}
```
Noted, thank you!

ivs-cetmix commented on setup/cetmix_tower/setup.py:
```diff
@@ -0,0 +1,6 @@
+import setuptools
+
```
This file is generated automatically

ivs-cetmix commented on cetmix_tower/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+# Copyright Cetmix OU
```
Noted, thank you!

ivs-cetmix committed:
[IMP] cetmix_tower_server: Command result

Replace 'COMMAND_RESULT' with 'result' in command 'code' field.
This is done to align with the general Odoo practice of using 'result'
and to keep the same coding style across the codebase.
Migration script is added to update existing commands.

Task: 4509

ivs-cetmix commented on cetmix_tower_server_notify_backend/README.rst:
```diff
@@ -68,6 +71,6 @@ Authors
 -----------
 
-This module is part of the `cetmix/cetmix-tower <https://github.com/cetmix/cetmix-tower/tree/14.0-dev/cetmix_tower_server_notify_backend>`_ project on GitHub.
+This module is part of the `cetmix/cetmix-tower <https://github.com/cetmix/cetmix-tower/tree/14.0/cetmix_tower_server_notify_backend>`_ project on GitHub.
```
Ok. But this file is generated automatically.

ivs-cetmix commented on cetmix_tower_yaml/readme/CONFIGURE.md:
```diff
@@ -1,5 +1 @@
-- Go to "Settings -> Users" and select a user.
-- Scroll to the "YAML export/import" section.
-- Activate "Allow YAML export" and/or "Allow YAML" import options.
+Please refer to the [official documentation](https://cetmix.com/tower) for detailed configuration instructions.
```
Thank you, we keep it up to date.

ivs-cetmix commented on cetmix_tower_yaml/static/description/index.html:
```diff
@@ -369,175 +369,55 @@ <h1 class="title">Cetmix Tower YAML</h1>
-<li><a class="reference internal" href="#credits" id="toc-entry-10">Credits</a><ul>
-<li><a class="reference internal" href="#authors" id="toc-entry-11">Authors</a></li>
-<li><a class="reference internal" href="#maintainers" id="toc-entry-12">Maintainers</a></li>
+<li><a class="reference internal" href="#configuration" id="toc-entry-1">Configuration</a></li>
+<li><a class="reference internal" href="#usage" id="toc-entry-2">Usage</a></li>
+<li><a class="reference internal" href="#bug-tracker" id="toc-entry-3">Bug Tracker</a></li>
+<li><a class="reference internal" href="#credits" id="toc-entry-4">Credits</a><ul>
+<li><a class="reference internal" href="#authors" id="toc-entry-5">Authors</a></li>
+<li><a class="reference internal" href="#maintainers" id="toc-entry-6">Maintainers</a></li>
 </ul>
```
This file is auto generated

ivs-cetmix commented on cetmix_tower_yaml/static/description/index.html:
```diff
@@ -369,175 +369,55 @@ <h1 class="title">Cetmix Tower YAML</h1>
 <div class="section" id="bug-tracker">
-<h1><a class="toc-backref" href="#toc-entry-9">Bug Tracker</a></h1>
+<h1><a class="toc-backref" href="#toc-entry-3">Bug Tracker</a></h1>
 <p>Bugs are tracked on <a class="reference external" href="https://github.com/cetmix/cetmix-tower/issues">GitHub Issues</a>.
 In case of trouble, please check there if your issue has already been reported.
 If you spotted it first, help us to smash it by providing a detailed and welcomed
-<a class="reference external" href="https://github.com/cetmix/cetmix-tower/issues/new?body=module:%20cetmix_tower_yaml%0Aversion:%2014.0-dev%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**">feedback</a>.</p>
+<a class="reference external" href="https://github.com/cetmix/cetmix-tower/issues/new?body=module:%20cetmix_tower_yaml%0Aversion:%2014.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**">feedback</a>.</p>
```
This file is auto generated. How many times should I repeat this?

ivs-cetmix commented on cetmix_tower_yaml/static/description/index.html:
```diff
@@ -369,175 +369,55 @@ <h1 class="title">Cetmix Tower YAML</h1>
-<li>File Templates</li>
-<li>Server Templates</li>
-</ul>
+<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Beta" src="https://img.shields.io/badge/maturity-Beta-yellow.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/licence-AGPL--3-blue.png" /></a> <a class= ...
+<p>This module implements YAML format data import/export for <a class="reference external" href="https://cetmix.com/tower">Cetmix
+Tower</a>.</p>
+<p>Please refer to the <a class="reference external" href="https://cetmix.com/tower">official
+documentation</a> for detailed information.</p>
```
This file is auto generated. How many times should I repeat this?

ivs-cetmix commented on cetmix_tower/static/description/index.html:
```diff
@@ -0,0 +1,620 @@
+        <div class="container">
+            <div class="row">
+                <div class="col-lg-8">
+                  <h2 style="color: #0D3045; font-weight: 500; font-size: 48px" data-bs-original-title="" title="" aria-describedby="tooltip483440">
+                    <strong style="font-weight: 800; color: #29A1DC; vertical-align: text-bottom;">| </strong>  What is Cetmix Tower?
+                      <br/>
+                    </h2>
```
Ok, noted

ivs-cetmix commented on cetmix_tower_server_queue/README.rst:
```diff
@@ -66,6 +71,6 @@ Authors
 -----------
 
-This module is part of the `cetmix/cetmix-tower <https://github.com/cetmix/cetmix-tower/tree/14.0-dev/cetmix_tower_server_queue>`_ project on GitHub.
+This module is part of the `cetmix/cetmix-tower <https://github.com/cetmix/cetmix-tower/tree/14.0/cetmix_tower_server_queue>`_ project on GitHub.
```
It's fine

ivs-cetmix commented on cetmix_tower_server_notify_backend/README.rst:
```diff
@@ -17,15 +17,19 @@ Cetmix Tower Server Notify Backend
     :alt: License: AGPL-3
 .. |badge3| image:: https://img.shields.io/badge/github-cetmix%2Fcetmix--tower-lightgray.png?logo=github
-    :target: https://github.com/cetmix/cetmix-tower/tree/14.0-dev/cetmix_tower_server_notify_backend
+    :target: https://github.com/cetmix/cetmix-tower/tree/14.0/cetmix_tower_server_notify_backend
```
This file is auto generated. How many times should I repeat this?

ivs-cetmix commented on cetmix_tower_server_notify_backend/README.rst:
```diff
@@ -35,25 +39,24 @@ https://cetmix.com/tower
 In case of trouble, please check there if your issue has already been reported.
 If you spotted it first, help us to smash it by providing a detailed and welcomed
-`feedback <https://github.com/cetmix/cetmix-tower/issues/new?body=module:%20cetmix_tower_server_notify_backend%0Aversion:%2014.0-dev%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+`feedback <https://github.com/cetmix/cetmix-tower/issues/new?body=module:%20cetmix_tower_server_notify_backend%0Aversion:%2014.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
This file is auto generated. How many times should I repeat this?

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -942,8 +891,75 @@ def run_command(self, command, path=None, sudo=None, ssh_connection=None, **kwar
+        # Get variables from path
+        path = path if path else command.path
+        if path:
+            variables_extracted = command.get_variables_from_code(path)
+            for ve in variables_extracted:
+                if ve not in variables:
+                    variables.append(ve)
+
+        # Get variable values for current server
+        variable_values_dict = (
+            self.sudo().get_variable_values(variables)  # pylint: disable=no-member
+            if variables
+            else False
+        )
+
+        # Extract variable values for current server
+        variable_values = (
+            variable_values_dict.get(self.id) if variable_values_dict else False
+        )  # pylint: disable=no-member
+
+        # Render command code using variables
+        if variable_values:
+            if command.action == "python_code":
+                variable_values["pythonic_mode"] = True
+
+            rendered_code = (
+                command.render_code_custom(command.code, **variable_values)
+                if command.code
+                else False
+            )
+            rendered_path = (
+                command.render_code_custom(path, **variable_values) if path else False
+            )
+
+        else:
+            rendered_code = command.code
+            rendered_path = path
+
+        return {"rendered_code": rendered_code, "rendered_path": rendered_path}
+
```
Thank you! We will keep it in mind.

GabbasovDinar reviewed (approved):
Code LGTM

ivs-cetmix committed:
[IMP] cetmix_tower_server: command timeout

Check 'zombie' commands and mark them as finished.
This is needed to avoid infinite loops when a command is stuck.

Task: 4509

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -942,8 +891,75 @@ def run_command(self, command, path=None, sudo=None, ssh_connection=None, **kwar
+        # Get variables from path
+        path = path if path else command.path
+        if path:
+            variables_extracted = command.get_variables_from_code(path)
+            for ve in variables_extracted:
+                if ve not in variables:
+                    variables.append(ve)
+
+        # Get variable values for current server
+        variable_values_dict = (
+            self.sudo().get_variable_values(variables)  # pylint: disable=no-member
+            if variables
+            else False
+        )
+
+        # Extract variable values for current server
+        variable_values = (
+            variable_values_dict.get(self.id) if variable_values_dict else False
+        )  # pylint: disable=no-member
+
+        # Render command code using variables
+        if variable_values:
+            if command.action == "python_code":
+                variable_values["pythonic_mode"] = True
+
+            rendered_code = (
+                command.render_code_custom(command.code, **variable_values)
+                if command.code
+                else False
+            )
+            rendered_path = (
+                command.render_code_custom(path, **variable_values) if path else False
+            )
+
+        else:
+            rendered_code = command.code
+            rendered_path = path
+
+        return {"rendered_code": rendered_code, "rendered_path": rendered_path}
+
```
Noted, will do it later

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command_log.py:
```diff
@@ -132,25 +150,24 @@ def finish(
 
-            # Trigger post finish hook
-            rec._command_finished()
+        # Trigger post finish hook
+        self_with_sudo._command_finished()
```
This is done with the "cetmix_tower_server_queue" module

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command_log.py:
```diff
@@ -28,10 +30,15 @@ class CxTowerCommandLog(models.Model):
     )
     # -- Command
-    is_running = fields.Boolean(help="Command is being executed right now")
+    is_running = fields.Boolean(
+        help="Command is being executed right now",
+        compute="_compute_duration",
+        store=True,
+    )
```
Sure, this is done

ivs-cetmix commented on cetmix_tower_server/migrations/14.0.0.6.2 copy/post-migration.py:
```diff
@@ -0,0 +1,22 @@
+    """
+    )
+
+    _logger.warning("-== Migration completed ==-")
```
Thank you for noticing this!

ivs-cetmix committed:
Fix directory structure

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -762,71 +684,101 @@ def test_ssh_connection(
+            Char: password ready to be used for connection parameters
         """
         self.ensure_one()
+        password = self.ssh_password
```
should we return password with sudo? 

`self.sudo().ssh_password`

ivs-cetmix replied:
Probably yes, thank you for noticing this!

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -762,71 +684,101 @@ def test_ssh_connection(
-            for ve in variables_extracted:
-                if ve not in variables:
-                    variables.append(ve)
+    @after_commit
```
perhaps we can also rename this decorator to a more understandable one e.g.: 

`@with_ssh_cleanup`

what do you think?

ivs-cetmix replied:
Sounds like a good idea!

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1634,45 +1651,35 @@ def _parse_command_results(
+        timeout = int(
+            self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("cetmix_tower_server.command_timeout", 0)
```
this `command_timeout` used only for zombie commands? maybe for any command without timeout we can use this parameter ...

ivs-cetmix replied:
Probably we can do this is the future because currently we don't have a direct method to control command timeout.

ivs-cetmix commented on cetmix_tower_server/models/res_config_settings.py:
```diff
@@ -9,7 +9,14 @@ class ResConfigSettings(models.TransientModel):
     _inherit = "res.config.settings"
 
-    def action_configure_cron(self):
+    cetmix_tower_command_timeout = fields.Integer(
+        string="Command Timeout",
+        config_parameter="cetmix_tower_server.command_timeout",
+        help="Timeout for commands in seconds after which"
+        " the command will be terminated",
+    )
```
default key value is not stored in the database, so it not returned by the get_config_param function.
So we adding the default value directly in the database using /data/ir_config_parameter.xml file
0 means "do not use timeout". No need to raise an exception for this.

ivs-cetmix commented on cetmix_tower_server/models/res_config_settings.py:
```diff
@@ -20,7 +27,24 @@ def action_configure_cron(self):
+            "target": "new",
+        }
+
+    def action_configure_zombie_commands_cron(self):
+        """
+        Configure cron job to check zombie commands
+        """
+        self.ensure_one()
+        cron_id = self.env.ref("cetmix_tower_server.ir_cron_check_zombie_commands").id
+        if not cron_id:
+            raise ValidationError(_("Cron job not found"))
+        return {
+            "name": _("Cron Job"),
```
No need to do this

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_command_log.py:
```diff
@@ -28,10 +30,15 @@ class CxTowerCommandLog(models.Model):
     )
     # -- Command
-    is_running = fields.Boolean(help="Command is being executed right now")
+    is_running = fields.Boolean(
+        help="Command is being executed right now",
+        compute="_compute_duration",
+        store=True,
+    )
```
Readonly is True by default for computed fiedls

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cetmix/cetmix-tower?pullRequest=240) <br/>All committers have signed the CLA.

Bearnard21 reviewed (approved):
LGTM

ivs-cetmix commented:
/ocabot merge nobump

CetmixGitDrone commented:
Hey, thanks for contributing! Proceeding to merge this for you.
Prepared branch [14.0-dev-ocabot-merge-pr-240-by-ivs-cetmix-bump-nobump](https://github.com/cetmix/cetmix-tower/commits/14.0-dev-ocabot-merge-pr-240-by-ivs-cetmix-bump-nobump), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at d1b4d9a47ddf5bfd8fd36b111a9e56f4aaaa8416. Thanks a lot for contributing to cetmix. ❤️
